Implement OpenCDM SDP prototype

ajutras-linaro opened the pull request:
Signed-off-by: Alexandre Jutras <alexandre.jutras@nxp.com>

petegriffin reviewed (commented):
Hi Alex,

Many thanks for the pull request. It would be good to have a more verbose commit message :-) mainly for other folks who maybe haven't attended our regular dev calls but are looking through the code.

Specifically: -
* A high level description of what opencdm sdp protoype is and does. e.g. allocates a secure buffer using ion, sends the descriptor via blah, relies on memory protection being disabled, tested on x platform.

It would also be good to document which version of ION api this commit is using (I guess "the old" ION api in v4.9 kernel?).

My other main comment would be it seems a very "c centric" solution to the problem. I'm thinking maybe a bufferallocator or buffer class maybe better, and then do operations such as ion->allocate(); ion->map() etc (I think we actually had similar functionality in the mediadrm playready plugin originally).

petegriffin commented on src/com/mediaengine/rpc/rpc_cdm_mediaengine_handler.cc:
```diff
@@ -50,6 +71,170 @@ RpcCdmMediaengineHandler::RpcCdmMediaengineHandler() {
+}
+
+
+static int copySecureData(int secure_fd, uint32_t secure_size, uint8_t *dst, uint32_t dst_size)
```
A comment would be good here to highlight that this should only be done as part of the protoype, and relies on memory protection being disabled of the secure buffers.

petegriffin commented on src/com/mediaengine/rpc/rpc_cdm_mediaengine_handler.cc:
```diff
@@ -19,6 +19,23 @@
+#include <sys/socket.h>
+#include <sys/ioctl.h>
+#include <sys/mman.h>
+#include "linux/ion.h"
```
Where is this header being included from?

ajutras-linaro replied:
ion.h is included from the intermediate file: tmp/sysroots/imx8mqevk/usr/include/linux/ion.h. This is copied from the kernel source: ./drivers/staging/android/uapi/ion.h. 

Is that ok?

Note that, since this is a system include, I will replace the "" by <> in the next pull request.

petegriffin commented on src/com/mediaengine/rpc/rpc_cdm_mediaengine_handler.cc:
```diff
@@ -19,6 +19,23 @@
+
+// TODO Organize a compile time flag
+// i.MX8M VPU secure heap
+#define ION_SECURE_HEAP_ID_DECODER 2
```
Would be good to do this "todo" so we can have it working on both hikey and iMX8M

petegriffin commented on src/com/mediaengine/rpc/rpc_cdm_mediaengine_handler.cc:
```diff
@@ -50,6 +71,170 @@ RpcCdmMediaengineHandler::RpcCdmMediaengineHandler() {
+  int ion;
+  int fd = -1;
+
+  ion = open("/dev/ion", O_RDWR);
```
We most likely need some way of detecting which ION API the kernel is exposing and either supporting both (best case) for the buffer allocation. Or alternatively erroring nicely if the kernel is newer than what we support.  I think xtest --sdp-basic unit test already has some method for trying to do this detection (although IIRC I think it was actually broken).

ajutras-linaro replied:
I would like to handle this comment separately. I have created a separate JIRA ticket (MMWG-48) for that purpose.

ajutras-linaro commented:
@petegriffin, thank you for your comments. I will take them into account.

psivasubramanian commented on src/com/mediaengine/rpc/rpc_cdm_mediaengine_handler.cc:
```diff
@@ -39,6 +56,10 @@ RpcCdmMediaengineHandler::RpcCdmMediaengineHandler() {
   }
 
+#ifdef OCDM_SDP_PROTOTYPE
+  mSocketFd = -1;
```
Can this be a enum like DEFAULT_SOCKETFD?

psivasubramanian commented on src/com/mediaengine/rpc/rpc_cdm_mediaengine_handler.cc:
```diff
@@ -50,6 +71,170 @@ RpcCdmMediaengineHandler::RpcCdmMediaengineHandler() {
+static int connectSocket(void)
+{
+  int status = 0;
+  int socket_fd = -1;
```
Replace with enum instead?

psivasubramanian commented on src/com/mediaengine/rpc/rpc_cdm_mediaengine_handler.cc:
```diff
@@ -100,11 +285,29 @@ bool RpcCdmMediaengineHandler::CreateMediaEngineSession(char *session_id_val,
+     at every decrypt operation. */
+  this->mSocketFd = connectSocket();
+  if(this->mSocketFd < 0) {
+    return false;
```
A error message here would be helpful to debug.

psivasubramanian commented on src/com/mediaengine/rpc/rpc_cdm_mediaengine_handler.cc:
```diff
@@ -50,6 +71,170 @@ RpcCdmMediaengineHandler::RpcCdmMediaengineHandler() {
+  if(status >= 0) {
+    status = 0;  // Success
+  } else {
+    if(errno == EWOULDBLOCK || errno == EAGAIN)
```
A error message here will be good for debugging POV.

psivasubramanian commented on src/com/mediaengine/rpc/rpc_cdm_mediaengine_handler.cc:
```diff
@@ -50,6 +71,170 @@ RpcCdmMediaengineHandler::RpcCdmMediaengineHandler() {
+    return -1;
+  }
+
+  memcpy(dst, secure_mem, secure_size);
```
what if dst is null and bytes copied is less than secure_size?

ajutras-linaro replied:
@psivasubramanian, what do you mean by "bytes copied is less than secure_size"? The size is already validated at line 215.

psivasubramanian commented on src/com/mediaengine/rpc/rpc_cdm_mediaengine_handler.cc:
```diff
@@ -50,6 +71,170 @@ RpcCdmMediaengineHandler::RpcCdmMediaengineHandler() {
+  int ion;
+  int fd = -1;
+
+  ion = open("/dev/ion", O_RDWR);
```
what about enum for device "/dev/ion"?

ajutras-linaro replied:
Do you mean to create a definition for "/dev/ion"? Like:

#define ION_DEVICE_NAME "/dev/ion"

psivasubramanian replied:
Exactly.

ajutras-linaro commented:
Thank you Siva for your comments. I will take them into account.

ajutras-linaro committed:
Communicate the socket channel ID when creating a media engine session

Modify the create media engine session RPC to communicate the socket channel
ID. This ID is used by the Secure Data Path implementation to uniquely create a
socket connection for each session. This is required to support multi-session
in the SDP implementation.

Signed-off-by: Alexandre Jutras <alexandre.jutras@linaro.org>

ajutras-linaro committed:
Remove unused variable

Signed-off-by: Alexandre Jutras <alexandre.jutras@linaro.org>

ajutras-linaro committed:
Implement an ION allocator helper class

This helper class is used to implement the SDP Prototype in the OpenCDM
browser plugin. It allows to allocate the 'secure' ION buffer used to output
the decrypted content. The buffer is allocated in the specified secure ION
heap ID. Once the decryption is completed, it allows to map the ION buffer
and to copy the data to the shared memory. This assumes that the 'secure'
ION heap is CPU accessible meaning it is not actually protected by any
hardware mechanism such as TZASC.

The current implementation only supports the ION ABI as defined in kernel 4.9.

Signed-off-by: Alexandre Jutras <alexandre.jutras@linaro.org>

ajutras-linaro committed:
Implement a socket client helper class

This socket client helper class allows the OpenCDM browser plugin to send the
secure (ION) file decriptor to the CDMI Service. The socket server helper class
implements the service side of the communication.

The helper class provides an API to connect to a socket (typically when
creating a media engine session) and to send the file descriptor (typically
at every decrypt operation). The file descriptor is communicated through the
socket's ancillary data. The secure buffer size is communicated through the
socket's data.

Signed-off-by: Alexandre Jutras <alexandre.jutras@linaro.org>

ajutras-linaro committed:
Implement the OpenCDM SDP Prototype

The purpose of the OpenCDM SDP Prototype is to validate the integration of the
ION buffer into OpenCDM and into the DRM layer. This is used to demonstrate
the OpenCDM's part of the SDP implementation on platforms which cannot
implement the end-to-end SDP or on platforms which does not support it yet.

The prototype is implemented at the OpenCDM browser plugin level. At every
decryption operation, it allocate a 'secure' ION buffer from the ION heap
specified by the ION_SECURE_HEAP_ID_DECODER compile-time variable. Then, it
communicates the corresponding file descriptor to the DRM layer via the CDMI
service. The interprocess communication relies on a per-session socket
connection established between the OpenCDM browser plugin and the CDMI service.

Once the DRM implementation has completed the content decryption into the
'secure' buffer, the control is returned to the browser plugin. It maps the
memory referenced by ION buffer, copies the data to the shared memory and
passes that shared memory to the audio and video frameworks to proceed with
the audio/video pipeline. This mechanism assumes that the 'secure' ION heap is
CPU accessible meaning it is not actually protected by any hardware mechanism
such as TZASC.

This has been tested on NXP i.MX8M using PlayReady's OpenCDM integration in
Chromium.

Signed-off-by: Alexandre Jutras <alexandre.jutras@linaro.org>

ajutras-linaro commented:
I am providing another set of patches to take your comments into account. Note that the current ION implementation only supports the ION ABI as defined in kernel 4.9. MMWG-48 will track that effort separately.

petegriffin commented:
Hi Alex, Thanks for the updated patches. Looks good to me. Peter/
